Simplify Message pump ack and reject methods

bording opened the pull request:
This is the set of changes originally in #165 related to the `MessagePump` class.

I've split them into two commits, so hopefully that makes it easier to follow the changes.

adamralph commented on src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs:
```diff
@@ -174,74 +174,53 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
             }
+            catch (AlreadyClosedException ex)
+            {
+                Logger.Warn($"Attempt to acknowledge/reject message {messageId} failed because the channel was closed. The message will be requeued.", ex);
```
Does it matter that the log message has changed from explicitly say "acknowledge" or "reject" depending on which operation was attempted to "acknowledge/reject"?

If we care about this, it would be fairly simple to lift a version of this catch into each branch of the `if`.

adamralph replied:
...or set a variable which tells us which operation was attempted (but that feels kludgey).

bording replied:
I didn't do that originally since I'd need to add additional try blocks around each call, but you're right that preserving the ack/reject distinction in the log would be useful.

adamralph commented on src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs:
```diff
@@ -174,74 +174,53 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
+                {
+                    await RejectMessage(channel, message.DeliveryTag).ConfigureAwait(false);
+                }
+                catch (AlreadyClosedException) { }
```
After reading [your response to my initial question](https://github.com/Particular/NServiceBus.RabbitMQ/pull/165#discussion_r68841837) about this I now understand why it's OK to swallow this exception here.

However, in the old version of the code, we would have executed this in this case:

```C#
Logger.Warn($"Attempt to reject message {messageId} failed because the channel was closed. The message will be requeued.", ex);
```

Is it worth preserving that? We could also preserve the "rejected" instead of the "requeued" in the log message above.

bording replied:
Since we need the catch here anyway, adding the logging back seems like an ok thing to do.

adamralph commented on src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs:
```diff
@@ -201,26 +201,33 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
             {
-                channel.BasicAck(deliveryTag, false);
-            });
+                var s = (MessageState)messageState;
```
Silly nitpick, but as a way of avoiding `s`, I'd probably prefer `state` for the lambda param and `messageState` for the local.

adamralph commented:
@bording I've reviewed the changes and left a couple of comments.

It's a great simplification. The current incarnations of `AcknowledgeMessage()` and `RejectMessage()` are quite nasty.

bording commented:
> The current incarnations of AcknowledgeMessage() and RejectMessage() are quite nasty.

I agree. I was never happy with how they turned out initially.

I've pushed up some fixup commits to address your comments, and I also added some helper methods around using `Task.Factory.StartNew` that I think make using it a bit nicer. Let me know what you think!

I'll squash the fixups once you've reviewed the changes.

adamralph commented on src/NServiceBus.RabbitMQ/TaskEx.cs:
```diff
@@ -1,10 +1,20 @@
+
+        public static Task StartNew(object state, Func<object, Task> function) => StartNew(state, function, TaskScheduler.Default);
+
+        public static Task StartNew(object state, Func<object, Task> function, TaskScheduler scheduler) => Task.Factory.StartNew(function, state, CancellationToken.None, TaskCreationOptions.DenyChildAttach, scheduler).Unwrap();
```
I like the simplification this affords in `MessagePump` but I think only the 2nd overload should exist as it's the only one that's required. The 1st, 3rd and 4th overloads aren't used so I'd call YAGNI on them and remove them. If we find the further overloads are needed later it's trivial to add them.

Also, given the method is only used in `MessagePump`, I'd probably say YAGNI on it being internal and make it a private in that class.

BTW - I think this would also work:

```C#
static Task StartNew<T>(T state, Action<T> action, TaskScheduler scheduler) =>
    Task.Factory.StartNew(
        o => action((T)o), state, CancellationToken.None, TaskCreationOptions.DenyChildAttach, scheduler);
```

It saves having to do the cast in the delegate that's passed into the method.

bording replied:
While the generic version does make the casting nicer, it has a closure (over `action`) so it reintroduces a per-call allocation.

adamralph commented:
@bording the fix up commits look spot on :ok_hand: 

I left a comment on the new helper method commit.

bording commented:
I went ahead and removed the currently unused helper methods, but I decided to keep the one being used in the `TaskEx` class, because I know there are going to be other places I want to be able to call it once this is merged.

I've squashed the fixup commits. Are they any other commits you think should be combined?

adamralph commented:
Thanks @bording :ok_hand:
